pallets/providers/src/utils.rs

```diff
@@ -499,13 +505,296 @@ where
+        MainStorageProviders::<T>::insert(&msp_id, msp);
+
+        // Return the old capacity
+        Ok(old_capacity)
+    }
+
+    /// This function holds the logic that checks if a user can change its capacity as a Backup Storage Provider
+    /// and, if so, updates the storage to reflect the new capacity, modifying the user's deposit accordingly
+    /// and returning the old capacity if successful
+    pub fn do_change_capacity_bsp(
+        account_id: &T::AccountId,
+        bsp_id: BackupStorageProviderId<T>,
+        new_capacity: StorageData<T>,
+    ) -> Result<StorageData<T>, DispatchError> {
+        // Check that the BSP is registered and get its info
+        let mut bsp = BackupStorageProviders::<T>::get(&bsp_id).ok_or(Error::<T>::NotRegistered)?;
+
+        // Check that the new capacity is diferent from the current capacity
+        ensure!(
+            new_capacity != bsp.capacity,
+            Error::<T>::NewCapacityEqualsCurrentCapacity
+        );
+
+        // Check that enough time has passed since the last capacity change
+        ensure!(
+            frame_system::Pallet::<T>::block_number()
+                >= bsp.last_capacity_change + T::MinBlocksBetweenCapacityChanges::get(),
+            Error::<T>::NotEnoughTimePassed
+        );
+
+        // Check that the new capacity is bigger than the minimum required by the runtime
+        ensure!(
+            new_capacity >= T::SpMinCapacity::get(),
+            Error::<T>::StorageTooLow
+        );
+
+        // Check that the new capacity is bigger than the current used capacity by the BSP
+        ensure!(
+            new_capacity >= bsp.data_used,
+            Error::<T>::NewCapacityLessThanUsedStorage
+        );
+
+        // Calculate how much deposit will the signer have to pay to register with this amount of data
+        let capacity_over_minimum = new_capacity
+            .checked_sub(&T::SpMinCapacity::get())
+            .ok_or(Error::<T>::StorageTooLow)?;
+        let deposit_for_capacity_over_minimum = T::DepositPerData::get()
+            .checked_mul(&capacity_over_minimum.into())
+            .ok_or(DispatchError::Arithmetic(ArithmeticError::Overflow))?;
+        let new_deposit = T::SpMinDeposit::get()
+            .checked_add(&deposit_for_capacity_over_minimum)
+            .ok_or(DispatchError::Arithmetic(ArithmeticError::Overflow))?;
+
+        // Check how much has the used already deposited for the current capacity
+        let current_deposit = T::NativeBalance::balance_on_hold(
+            &HoldReason::StorageProviderDeposit.into(),
+            account_id,
+        );
+
+        // Check if the new deposit is bigger or smaller than the current deposit
+        // Note: we do not check directly capacities as, for example, a bigger new_capacity could entail a smaller deposit
+        // because of changes in storage pricing, so we check the difference in deposits instead
+        if new_deposit > current_deposit {
```

Comment:
Can we extract this logic of the `if` and `else` into functions like `hold_balance` and `release_balance`? Both functions `do_change_capacity_bsp` and `do_change_capacity_msp` are the same.

Reply:
Done!